manifests/instance.pp

```diff
@@ -0,0 +1,232 @@
+#     ports            => '8800',
+#     requires         => Package['haproxy'],
+#   }
+#   haproxy::instance { 'group2': }
+#   haproxy::instance_service { 'group2':
+#     haproxy_init_source => "puppet:///modules/${module_name}/haproxy-group1.init",
+#   }
+#   haproxy::listen { 'puppet00':
+#     instance         => 'group2',
+#     collect_exported => false,
+#     ipaddress        => $::ipaddress,
+#     ports            => '9900',
+#     requires         => Package['haproxy'],
+#   }
+#
+# Multiple instances of haproxy, one with a custom haproxy package:
+#   haproxy::instance { 'group1': }
+#   haproxy::instance_service { 'group1':
+#     haproxy_init_source => "puppet:///modules/${module_name}/haproxy-group1.init",
+#   }
+#   haproxy::listen { 'puppet00':
+#     instance         => 'group1',
+#     collect_exported => false,
+#     ipaddress        => $::ipaddress,
+#     ports            => '8800',
+#     requires         => Package['haproxy'],
+#   }
+#   haproxy::instance { 'group2': }
+#   haproxy::instance_service { 'group2':
+#     haproxy_package     => 'custom_haproxy',
+#     haproxy_init_source => "puppet:///modules/${module_name}/haproxy-group2.init",
+#   }
+#   haproxy::listen { 'puppet00':
+#     instance         => 'group2',
+#     collect_exported => false,
+#     ipaddress        => $::ipaddress,
+#     ports            => '9900',
+#     requires         => Package['haproxy'],
+#   }
+#
+#  When running multiple instances on one host, there must be a Service[] for
+#  each instance.  One way to create the situation where Service[] works is
+#  using haproxy::instance_service.
+#  However you may want to do it some other way. For example, you may
+#  not have packages for your custom haproxy binary. Or, you may wish
+#  to use the standard haproxy package but not create links to it, or
+#  you may have different init.d scripts.  In these cases, write your own
+#  puppet code that will result in Service[] working for you and do not
+#  call haproxy::instance_service.
+#
+define haproxy::instance (
+  $package_ensure   = 'present',
+  $package_name     = undef,
+  $service_ensure   = 'running',
+  $service_manage   = true,
+  $global_options   = undef,
+  $defaults_options = undef,
+  $restart_command  = undef,
+  $custom_fragment  = undef,
+  $config_file      = undef,
+) {
+
+  if $service_ensure != true and $service_ensure != false {
+    if ! ($service_ensure in [ 'running','stopped']) {
+      fail('service_ensure parameter must be running, stopped, true, or false')
+    }
+  }
+  validate_string($package_name,$package_ensure)
+  validate_bool($service_manage)
+
+  # Since this is a 'define', we can not use 'inherts haproxy::params'.
+  # Therefore, we "include haproxy::params" for any parameters we need.
+  if $global_options == undef {
+    include haproxy::params
+    $_global_options = $haproxy::params::global_options
+  } else {
+    $_global_options = $global_options
+  }
```

Comment:
If all of $global_options, $default_options and $config_file are defined, doing the "include haproxy::params" is a waste of time.  That said, I may be guilty of premature optimization.  Would you recommend I simplify the code and do one "include haproxy::params" early in the class?